Integrate common ISensor interface (net10.0)

framinosona opened the pull request:
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Description of Change

Extends sensor interfaces such as IAccelerometer, IBarometer, ICompass, IGyroscope, IMagnetometer, and IOrientationSensor with a common ISensor interface to unify capabilities. Eliminates redundant IsSupported and IsMonitoring properties and Start/Stop methods from individual sensor interfaces. Updates IAppActions, IHapticFeedback, IPhoneDialer, and IVibration to extend from IDeviceCapabilities for uniformity. Improves code maintainability by reducing duplicate functionality.

Relates to codebase cleanup and standardization.

(net9 PR : https://github.com/dotnet/maui/pull/30121)

jfversluis reviewed (changes requested):
84 errors that look similar to this

> C:\a\_work\1\s\src\Essentials\src\PublicAPI\net\PublicAPI.Shipped.txt(673,1): error RS0017: Symbol 'Microsoft.Maui.Devices.Sensors.IMagnetometer.Stop() -> void' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [C:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Essentials\src\PublicAPI\net\PublicAPI.Shipped.txt(675,1): error RS0017: Symbol 'Microsoft.Maui.Devices.Sensors.IOrientationSensor.IsMonitoring.get -> bool' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [C:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Essentials\src\PublicAPI\net\PublicAPI.Shipped.txt(676,1): error RS0017: Symbol 'Microsoft.Maui.Devices.Sensors.IOrientationSensor.IsSupported.get -> bool' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [C:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Essentials\src\PublicAPI\net\PublicAPI.Shipped.txt(678,1): error RS0017: Symbol 'Microsoft.Maui.Devices.Sensors.IOrientationSensor.Start(Microsoft.Maui.Devices.Sensors.SensorSpeed sensorSpeed) -> void' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [C:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Essentials\src\PublicAPI\net\PublicAPI.Shipped.txt(679,1): error RS0017: Symbol 'Microsoft.Maui.Devices.Sensors.IOrientationSensor.Stop() -> void' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [C:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net10.0]

framinosona commented:
@jfversluis Of course ... When I add them to Unshipped, I need to remove them in the Shipped ... I just pushed. It should fix it :)

framinosona commented:
/azp run MAUI-public

framinosona commented:
Ohhh hold on, just discovered there is *REMOVED* tag ! Will use that instead

rmarinho commented:
@framinosona  needs rebase please. Sorry.

framinosona committed:
refa(Sensors): Integrates device capability interfaces

Updates sensor and capability interfaces to extend common base interfaces
(ISensor and IDeviceCapabilities), promoting reusable design. Removes
redundant property and method declarations in specific sensor interfaces
by leveraging the base interfaces. Consolidates support detection and
monitoring state methods.

Improves code maintainability and enforces consistency across device
capability-related interfaces.

framinosona commented:
@rmarinho Done :)

rmarinho commented:
/azp run

rmarinho reviewed (changes requested):
Seems some api changes twice 

```
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt(26,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.HapticFeedback.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=netstandard2.0]
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt(27,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.Vibration.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=netstandard2.0]
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/net/PublicAPI.Unshipped.txt(28,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.HapticFeedback.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=net10.0]
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/net/PublicAPI.Unshipped.txt(29,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.Vibration.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=net10.0]
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt(26,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.HapticFeedback.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=netstandard2.1]
/Users/builder/azdo/_work/1/s/src/Essentials/src/PublicAPI/netstandard/PublicAPI.Unshipped.txt(27,1): error RS0025: The symbol 'static Microsoft.Maui.Devices.Vibration.IsSupported.get -> bool' appears more than once in the public API files (https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=netstandard2.1]
    0 Warning(s)
    6 Error(s)
```

framinosona committed:
refa (PublicAPI): Removes duplicated HapticFeedback and Vibration IsSupported

rmarinho commented:
/azp run

mattleibow reviewed (changes requested):
This is overall a good idea, however, I need to check with the language gurus if this is too much of a breaking change. For example, existing libraries using these interfaces need to still work.

Can you confirm that building a library using net8/9 `ICompass` will work without a recompile?

If not, maybe we can still have these interfaces, just with duplicate methods. Then the ICompass can implement the base interfaces explicitly.

For example:

```cs
using System;
public static class C {
    public static void Main() {
        var c = new Compass();

        IDeviceCap cap = c;
        ISensor sensor = c;

        Console.WriteLine("IsSupported: " + cap.IsSupported);
        Console.WriteLine("IsMonitoring: " + sensor.IsMonitoring);

        c.Start();
        Console.WriteLine("IsMonitoring: " + c.IsMonitoring);

        sensor.Stop();
        Console.WriteLine("IsMonitoring: " + c.IsMonitoring);
    }
}

class Compass : ICompass
{
    public bool IsSupported => true;
    public bool IsMonitoring { get;private set; }
    public void Start()
    {
        IsMonitoring = true;
    }
    public void Stop()
    {
        IsMonitoring = false;
    }
}

interface ICompass : ISensor
{
    new bool IsMonitoring { get; }
    new void Start();
    new void Stop();
    new bool IsSupported { get; }
    
    bool IDeviceCap.IsSupported => IsSupported;
    bool ISensor.IsMonitoring => IsMonitoring;
    void ISensor.Start() => Start();
    void ISensor.Stop() => Stop();
}

interface ISensor : IDeviceCap
{
    bool IsMonitoring { get; }
    void Start();
    void Stop();
}

interface IDeviceCap
{
    bool IsSupported { get; }
}
```

mattleibow commented:
I have been working on a PR with some non-breaking things: https://github.com/dotnet/maui/pull/31086

However, it is a bit unfortunate that it is a bit hacky.

Does this new set of interfaces help anything other that duplicate members on the interfaces? Not sure it is worth the breaks in this PR or the (maybe potential) issues in my PR.

For now, I am thinking not, but I am open to other opinions and thoughts.
